common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -22,64 +22,379 @@
+        }
+
+        public Slot getSlot() {
+            return slot;
+        }
+    }
 
     public ClientCertAuthChallengeHandler(@NonNull final Activity activity) {
         mActivity = activity;
-        //Create and start YubiKitManager
+        synchronized (deviceLock) {
+            //set mDevice to null
+            mDevice = null;
+        }
+
+        //Set mCurrentDialog to null
+        synchronized (smartcardDialogLock) {
+            mCurrentDialog = null;
+        }
+        //Create and start YubiKitManager for UsbDiscovery mode.
+        //When in Usb Discovery mode, Yubikeys that plug into the device will be accessible
+        // once the user provides permission via the Android permission dialog.
         mYubiKitManager = new YubiKitManager(mActivity.getApplicationContext());
         mYubiKitManager.startUsbDiscovery(new UsbConfiguration(), new Callback<UsbYubiKeyDevice>() {
             @Override
             public void invoke(UsbYubiKeyDevice device) {
                 Logger.info(TAG, "A YubiKey device was connected");
-                device.setOnClosed(new Runnable() {
-                    @Override
-                    public void run() {
-                        Logger.info(TAG, "A YubiKey device was disconnected");
-                    }
-                });
+                synchronized (deviceLock) {
+                    //set device
+                    mDevice = device;
+                    mDevice.setOnClosed(new Runnable() {
+                        @Override
+                        public void run() {
+                            synchronized (deviceLock) {
+                                Logger.info(TAG, "A YubiKey device was disconnected");
+                                mDevice = null;
+                                //Call onCancel on current dialog, if dialog is showing.
+                                synchronized (smartcardDialogLock) {
+                                    if (mCurrentDialog != null) {
+                                        mCurrentDialog.onCancelCba();
+                                        //Reset dialog to null
+                                        mCurrentDialog = null;
+                                    }
+                                }
+                            }
+                        }
+                    });
+                }
             }
         });
     }
 
+    //Called when a ClientCertRequest is received by the WebViewClient.
+    //Offers the user to choose a certificate to authenticate with based on whether or not a YubiKey is plugged in and has permission to be connected.
     @TargetApi(Build.VERSION_CODES.LOLLIPOP)
     @Override
     public Void processChallenge(@NonNull final ClientCertRequest request) {
-        //final String methodTag = TAG + ":processChallenge";
+        final String methodTag = TAG + ":processChallenge";
+        // Check if a YubiKey device has been discovered and has permission to be connected.
```

Comment:
Wondering if we should make mDevice a local variable.

If that can be done, then we don't have to worry about synchronization. We created one object and pass it all the way through the flow (as a final variable). The object will not be re-assigned.

Can we try extracting the mDevice initialization block in the constructor, and turn it into a synchronous function (with resultfuture, with a small timeout)?

Reply:
(not sure if possible since i don't know if processChallenge() is executed on main thread, and startUsbDiscovery() requires main thread to be unblocked.)

Reply:
So it sounds like we're trying  to find a way to detect the `UsbYubiKeyDevice` at the beginning of process challenge.
It's actually possible to call `startUsbDiscovery()` within `processChallenge` now, as I learned through testing. (I think the lack of `configChanges` keyboard flags were the culprit when I attempted this at the beginning of development.) I attempted moving the `startUsbDiscovery()` block to the beginning of `processChallenge`, creating a `ResultFuture` variable, calling `get` on the variable after the `startUsbDiscovery()` block, and calling `setResults` within the `startUsbDiscovery()` callback. 
However, `ResultFuture.get()` called within processChallenge does seem to block the callback for `startUsbDiscovery`, so I think both are relying on the same thread.